### ISSUE TITLE
Output all possible Texts of synz

### ISSUE THREAD

**Issue by TimonFelske** [RosaeNLG/rosaenlg#2]
Hello it's me again :)

I have a question regarding the synz and syn structure in rosaenlg.
I'd like to run some kind of debug-mode, to output all possible texts, that could be generated.
When using the synz with multiple syn texts, it always picks one of them, like mentioned in the reference documentation.

Is it possible, to output all the possible texts inside a synz with only one execution of the pug file?
Example:

    synz
     syn
      | the iphone is really a fantastic phone.
     syn
      | i really love the new iphone.
     syn
      | iphone  : a great phone!

With that code, I would like to output something like: "the iphone is really a fantastic phone./i really love the new iphone./iphone  : a great phone!" with just one execution.

If that is not possible, I could execute the pug file multiple times, until all different texts are used.
But I think it would be a nice feature to get all the possible texts in one render call, especially while debugging the program.

Best regards
Timon

**Comment by ludans** [comment]
Hello Timon,
Just saw the ticket.
Nice idea, I'll think about it.

**Comment by ludans** [comment]
When the synonymic alternatives are just text, it would be fairly easy to list them all using a separator like you proposed.
There is a (big) issue when alternatives contain code, or other synonyms: it seems impossible to use that approach.

The options I see:

1/ Generate a tree of all texts
As, eventually, it is a tree.
Drawbacks:
- the tree can be huge
- would it be readable?
- not mundane to implement

2/ Generate all the texts
A batch that would generate all the possible texts.
The output would be readable: each output is a possible text.
Drawbacks:
- there could be millions of texts; what's the goal of having them all?
- also not mundane to implement


My general advice, if synonyms along testability is an issue:
- avoid using synonyms, or use them less: this is often the case for financial applications when output must be completely predictable
- use regression testing and test some of the outputs:
	- some with a fixed random seed: must be an exact match
	- some with random seed, but checking for invariants in the text (typically numbers)

@TimonFelske any thoughts? What are the practical issues you are facing?

**Comment by TimonFelske** [comment]
I used the option I wrote, which is similar to your second option.
I used a loop to call the renderFile method multiple times.
There are only a few synonyms in my texts, so 30 iterations generated all possible texts, in my case.
I stored all the texts in a list and printed it after the 30 iterations.

My issue was, that I would like to check, if the texts that will be generated are all correct.
The texts will be used in a rehabilitation therapy with real patients. Furthermore the texts contain patient-related information.
Because of that, all the texts have to be checked, if they are correct and doesn't contain false information.
So it's just a feature to debug the texts.

**Comment by ludans** [comment]
I will try something on option 2 - a batch that would generate all the possible texts.
Probably in December.

**Comment by ludans** [comment]
I have tried to make something clever but did not succeed :( - overly complex, and the approach does not work.

I think your approach is, as of today, the best one.

I added some documentation to suggest a workaround, which is the same as what you did (execute the pug file multiple times), but with one improvement: changing `forceRandomSeed` parameter each time (put 1 then 2 then 3 etc.) which should be more repeatable than pure random.

https://github.com/RosaeNLG/rosaenlg/pull/10/files
